gpbft/participant.go

```diff
@@ -66,63 +68,62 @@ func (p *Participant) Start() (err error) {
 }
 
 func (p *Participant) CurrentRound() uint64 {
-	if p.granite == nil {
+	if p.gpbft == nil {
 		return 0
 	}
-	return p.granite.round
+	return p.gpbft.round
 }
 
-// Validates a message received from another participant, if possible.
-// An invalid message can never become valid, so may be dropped.
-// A message can only be validated if it is for the currently-executing protocol instance.
-// Returns whether the message could be validated, and an error if it was invalid.
-func (p *Participant) ValidateMessage(msg *GMessage) (checked bool, err error) {
+// Receives a Granite message from some other participant.
+//   - A message for a previous instance is dropped.
+//   - A message for the current instance is validated and delivered immediately.
+//   - A message for a future instance for which a committee is available is validated and
+//     locally queued, to be delivered when that instance begins.
+//   - A message for a future instance for which a committee is not available returns an error.
+//   - An invalid message returns an error.
+//
+// Returns whether the message was accepted for the current instance or a near future instance,
+// and/or an error if it was invalid or could not be processed.
+func (p *Participant) ReceiveMessage(msg *GMessage) (accepted bool, err error) {
```

Comment:
I have a slight issue with API here.
1. We know we need to leverage multithreading for signature validation. Otherwise, validating the signatures can be a bottleneck.
2. Participant does not handle concurrency
3. libp2p/pubsub validation by default is multithreaded

We could do validation from outside by fetching comities and calling `ValidateMessage,` but this validation is duplicated.

Reply:
Could you expand further please, Kuba?

If signature validation is the only pace we would need parallelism, would it be sufficient to encapsulate it within `Signer` interface?

Reply:
The validation from pubsub is multithreaded (and I would like to keep it like this). We could verify signatures separately and then take a lock to validate the rest, but at that point, we might as well make the whole validation thread safe.

Reply:
Summary of discussion with @Kubuxu at standup: 

There are three separate issues:
1) the need for validation as a self-contained functionality which may or may not result in a consecutive call to receive message.
2) avoiding re-validation of messages that are already validated.
3) parallel validation of messages in terms of performance concerns.

Regarding the first issue, GossipSub has a separate validate flow and handle flow, where a client provides a validation hook that is called by GossipSub upon receiving messages and is required to be safe for parallel execution. A separate validate flow in GossipSub is there to reduce the possibility of accidentally spamming the network, and only propagate messages that have at least passed validation of honest nodes. Therefore, we need some self-contained way of offering that functionality in go-f3.

Regarding the second issue, although implicit message validation by the `Receiver` interface in F3 is reasonable defensive programming, it means we will end up with messages getting validated twice: once by the gossipSub validation mechanism mentioned earlier and once internally by F3 implementation. This may be OK if validation is cheap? but ideally we should offer a way avoid this.

Regarding the third issue, we should make message validation as fast to execute as possible while keeping it safe for parallel execution. This can be tricky since validation at certain steps depends on internal state of the participant such as instance ID, phase, etc.

Solutions discussed:
* _Revert back to the previous API_: The previous gPBFT API, facilitated conditional validation on receive message. This meant that the GossipSub integration layer could signal to the receiver that a message is already validated and the implementation may skip validation. But this can be error prone.
* _Introduce dedicated `Validator` interface that returns `ValidatedMessage` type which is then taken by the `Receiver`_: this could work but might be too much typing/combersome to work with.
* _Unexported internal struct fields that can signal conditional validation_: This would work but unclear where it leaves the exported gPBFT API.

This API change blocks the gossipSub integration and we should resolve this soon. An alternative there could be to take the first solution above and revisit once both this PR and gossipSub integration are landed. Then revisit.

@Kubuxu Please let me know if there is anything that I might have missed from our discussion. 🙏

Reply:
Thanks for the discussion. Validation is not cheap: validating aggregate signatures on message justification is probably the most CPU-expensive thing in the whole protocol. I understand the need to support parallelism and will adjust the API to support this. I'll try the `Validator`/`ValidatedMessage` approach.